File: Stack/Python/Next Greater Element II.py
# Problem link: https://leetcode.com/problems/next-greater-element-ii/description/
# Time: O(n)
# Space: O(n)
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        result = [-1] * len(nums)
        stack = []

        for idx in range(len(nums) * 2 - 1, -1, -1):
            while stack and stack[-1] <= nums[idx % len(nums)]:
                stack.pop()
            
            if idx < len(nums) and stack:
                result[idx] = stack[-1]
            
            stack.append(nums[idx % len(nums)])
        
        return result

File: Stack/Java/Next Greater Element II.java
// Problem link: https://leetcode.com/problems/next-greater-element-ii/description/ 
// Time: O(n)
// Space: O(n)
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int size = nums.length;
        int[] result = new int[size];
        Stack<Integer> stack = new Stack<>();

        for(int idx = size * 2 - 1; idx >= 0; idx--) {

            while(!stack.isEmpty() && stack.peek() <= nums[idx % size]) {
                stack.pop();
            }

            if(idx < size) {
                result[idx] = !stack.isEmpty() ? stack.peek() : -1;
            }

            stack.push(nums[idx % size]);
        }

        return result;
    }
}
